src/coprocessor/xeval/builtin_cast.rs

```diff
@@ -0,0 +1,392 @@
+        // TODO: add impl
+        Err(Error::Eval(ERROR_UNIMPLEMENTED.to_owned()))
+    }
+
+    pub fn cast_string_as_duration(&mut self, _ctx: &EvalContext, _expr: &Expr) -> Result<Datum> {
+        // TODO: add impl
+        Err(Error::Eval(ERROR_UNIMPLEMENTED.to_owned()))
+    }
+
+    pub fn cast_time_as_int(&mut self, _ctx: &EvalContext, _expr: &Expr) -> Result<Datum> {
+        // TODO: add impl
+        Err(Error::Eval(ERROR_UNIMPLEMENTED.to_owned()))
+    }
+
+    pub fn cast_time_as_real(&mut self, _ctx: &EvalContext, _expr: &Expr) -> Result<Datum> {
+        // TODO: add impl
+        Err(Error::Eval(ERROR_UNIMPLEMENTED.to_owned()))
+    }
+
+    pub fn cast_time_as_string(&mut self, _ctx: &EvalContext, _expr: &Expr) -> Result<Datum> {
+        // TODO: add impl
+        Err(Error::Eval(ERROR_UNIMPLEMENTED.to_owned()))
+    }
+
+    pub fn cast_time_as_decimal(&mut self, _ctx: &EvalContext, _expr: &Expr) -> Result<Datum> {
+        // TODO: add impl
+        Err(Error::Eval(ERROR_UNIMPLEMENTED.to_owned()))
+    }
+
+    pub fn cast_time_as_time(&mut self, _ctx: &EvalContext, _expr: &Expr) -> Result<Datum> {
+        // TODO: add impl
+        Err(Error::Eval(ERROR_UNIMPLEMENTED.to_owned()))
+    }
+
+    pub fn cast_time_as_duration(&mut self, _ctx: &EvalContext, _expr: &Expr) -> Result<Datum> {
+        // TODO: add impl
+        Err(Error::Eval(ERROR_UNIMPLEMENTED.to_owned()))
+    }
+
+    pub fn cast_duration_as_int(&mut self, ctx: &EvalContext, expr: &Expr) -> Result<Datum> {
+        let child = try!(self.get_one_child(expr));
+        let d = try!(self.eval(ctx, child));
+        if let Datum::Dur(_) = d {
+            let i = try!(try!(d.into_dec()).as_i64().into_result());
+            return Ok(Datum::I64(i));
+        }
+        invalid_type_error(&d, TYPE_DURATION)
+    }
+
+    pub fn cast_duration_as_real(&mut self, ctx: &EvalContext, expr: &Expr) -> Result<Datum> {
+        let child = try!(self.get_one_child(expr));
+        let d = try!(self.eval(ctx, child));
+        if let Datum::Dur(_) = d {
+            return Ok(Datum::F64(try!(try!(d.into_dec()).as_f64())));
+        }
+        invalid_type_error(&d, TYPE_DURATION)
+    }
+
+    pub fn cast_duration_as_string(&mut self, ctx: &EvalContext, expr: &Expr) -> Result<Datum> {
+        let child = try!(self.get_one_child(expr));
+        let d = try!(self.eval(ctx, child));
+        if let Datum::Dur(_) = d {
+            let s = try!(d.into_string());
+            return Ok(Datum::Bytes(s.into_bytes()));
+        }
+        invalid_type_error(&d, TYPE_DURATION)
+    }
+
+    pub fn cast_duration_as_decimal(&mut self, ctx: &EvalContext, expr: &Expr) -> Result<Datum> {
+        let child = try!(self.get_one_child(expr));
+        let d = try!(self.eval(ctx, child));
+        if let Datum::Dur(_) = d {
+            let dec = try!(d.into_dec());
+            return Ok(Datum::Dec(dec));
```

Comment:
add the call to `ProduceDecWithSpecifiedTp`, refer to
https://github.com/pingcap/tidb/blob/master/expression/builtin_cast.go#L877